core/drivers/imx_i2c.c

```diff
@@ -0,0 +1,392 @@
+
+	/* Enable the interface */
+	i2c_io_write8(bid, I2CR, I2CR_IEN);
+
+	tmp = i2c_io_read8(bid, I2CR) | I2CR_MSTA;
+	i2c_io_write8(bid, I2CR, tmp);
+
+	/* Wait until the bus is active */
+	ret = i2c_sync_bus(bid, &bus_is_busy, NULL);
+	if (ret != TEE_SUCCESS)
+		return ret;
+
+	/* Slave address on the bus */
+	return i2c_write_data(bid, &chip, 1);
+}
+
+TEE_Result imx_i2c_read(uint8_t bid, uint8_t chip, uint8_t *buf, int len)
+{
+	TEE_Result ret = TEE_SUCCESS;
+
+	if (bid >= ARRAY_SIZE(i2c_bus))
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	if ((len && !buf) || chip > 0x7F)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	ret = i2c_init_transfer(bid, chip << 1 | BIT(0));
+	if (ret == TEE_SUCCESS)
+		ret = i2c_read_data(bid, buf, len);
+
+	if (i2c_idle_bus(bid) != TEE_SUCCESS)
+		IMSG("bus not idle");
+
+	return ret;
+}
+
+TEE_Result imx_i2c_write(uint8_t bid, uint8_t chip, const uint8_t *buf, int len)
+{
+	TEE_Result ret = TEE_SUCCESS;
+
+	if (bid >= ARRAY_SIZE(i2c_bus))
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	if ((len && !buf) || chip > 0x7F)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	ret = i2c_init_transfer(bid, chip << 1);
+	if (ret == TEE_SUCCESS)
+		ret = i2c_write_data(bid, buf, len);
+
+	if (i2c_idle_bus(bid) != TEE_SUCCESS)
+		IMSG("bus not idle");
+
+	return ret;
+}
+
+TEE_Result imx_i2c_probe(uint8_t bid, uint8_t chip)
+{
+	if (bid >= ARRAY_SIZE(i2c_bus))
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	if (chip > 0x7F)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	return imx_i2c_write(bid, chip, NULL, 0);
+}
+
+/*
+ * I2C bus initialization: configure the IOMUX and enable the clock.
+ * @bid: Bus ID: (0=I2C1), (1=I2C2), (2=I2C3).
+ * @bps: Bus baud rate, in bits per second.
+ */
+TEE_Result imx_i2c_init(uint8_t bid, int bps)
+{
+	struct imx_i2c_mux *mux = &i2c_mux;
+
+	if (bid >= ARRAY_SIZE(i2c_bus))
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	if (!bps)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	io_write32(mux->base.va + mux->i2c[bid].scl_mux, I2C_MUX_VAL(bid));
+	io_write32(mux->base.va + mux->i2c[bid].scl_cfg, I2C_CFG_VAL(bid));
+	io_write32(mux->base.va + mux->i2c[bid].sda_mux, I2C_MUX_VAL(bid));
+	io_write32(mux->base.va + mux->i2c[bid].sda_cfg, I2C_CFG_VAL(bid));
+
+	/* Baud rate in bits per second */
+	i2c_set_bus_speed(bid, bps);
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result get_va(paddr_t pa, vaddr_t *va)
+{
+	if (!core_mmu_add_mapping(MEM_AREA_IO_SEC, pa, 0x10000))
+		return TEE_ERROR_GENERIC;
+
+	*va = (vaddr_t)phys_to_virt(pa, MEM_AREA_IO_SEC);
+	if (*va)
+		return TEE_SUCCESS;
+
+	return TEE_ERROR_GENERIC;
+}
+
+static TEE_Result i2c_early_init(void)
+{
+	size_t n = 0;
+
+	if (get_va(i2c_clk.base.pa, &i2c_clk.base.va) != TEE_SUCCESS)
+		return TEE_ERROR_GENERIC;
+
+	if (get_va(i2c_mux.base.pa, &i2c_mux.base.va) != TEE_SUCCESS)
+		return TEE_ERROR_GENERIC;
+
+	for (n = 0; n < ARRAY_SIZE(i2c_bus); n++) {
+		if (get_va(i2c_bus[n].pa, &i2c_bus[n].va) != TEE_SUCCESS)
+			return TEE_ERROR_GENERIC;
+	}
+
+	return TEE_SUCCESS;
+}
+
+driver_init(i2c_early_init);
```

Comment:
Minor: avoid confusion
Since there is now a `early_init()` registration macro and here you use `driver_init()`, maybe rename `i2c_early_init()`  to `i2c_init()` or `i2c_driver_init()`.

Reply:
yes, doing this. thanks